### ISSUE TITLE
Discussion about this cool library!

### ISSUE THREAD

**Issue by nbren12** [pangeo-data/xbatcher#2]
Hi Ryan,

@jhamman pointed me to this project, and it definitely seems like something that would be useful for the neural network + xarray community. Thanks for making it publicly available. 

Since I have been working in this space the last year, I have found it very difficult to find the right abstraction level for this kind of batch generation. For my part, I have been using PyTorch instead of the keras+tensorflow, and have been plugging in with their `Dataset` and `DataLoader` framework: https://pytorch.org/tutorials/beginner/data_loading_tutorial.html. In this framework, you write a `Dataset` object with a `__getitem__` method that returns a single "sample" (e.g. a pair of points, a patch of an image) etc, than the built in `DataLoader` handles all the work of combining these samples into batches. The advantage of this compared to plain iterators is that it allows an easy way to randomly shuffle the samples before combining them into batches. In my experience, this shuffling is basically required for a NN to train properly. FWIW, here is the `Dataset` object I have written for my work: https://github.com/nbren12/uwnet/blob/master/uwnet/datasets.py.

Overall, I think this is a challenging problem because the random access patterns we us in NN training may have nothing to do with how the data are laid out in memory or on disk.

**Comment by rabernat** [comment]
Noah -- thanks for your message and your suggestions.  I know you've been at it for a long time and have come very far. I am very new to it all. This project is a really rough sketch which I threw together quite quickly. We are just experimenting at this point.

If it turns out there is actually a need for a package like this, would love to collaborate with people like you who have already come up with their own solutions on specific datasets.

**Comment by nbren12** [comment]
Hey no problem! I think there is definitely a need for this sort of thing. Let me know if you ever want chat.

**Comment by tjvandal** [comment]
Hi @rabernat and @nbren12
 
I’ve been having discussions with @jhamman about this problem of loading data into pytorch/tensorflow from xarray. I’ll focus on Pytorch for now. My current process is to preprocess data from xarray into sample data files that can be easily loaded by pytorch’s `Dataset`. At scale this is a problem, I’m consistently way over data quotas on the HPC system I use. Often I can only use a subset of the total available dataset. Like you point out Noah, random access of this data is a requirement for training. 
 
There are a few ideas I’ve been thinking about here. Pytorch’s `Dataloader` already tries to preload the data onto the GPU so it seems that speeding up random sampling memory access should help. I’m wondering if there are some interesting ways to setup data chunks and sampling. Can chunks be defined as sets of training samples? It’s possible that generating intermediate training files is good enough and in that case an xarray to standard pytorch data format could be very helpful. I’m beginning to work with many different satellite/climate datasets so I’d be interested in thinking about this problem more.

**Comment by nbren12** [comment]
Hi @tjvandal. I really enjoyed your talk at the NOAA AI workshop last spring, btw.

We really have developed some best practices around this that I think works pretty well. Kudos go mostly to Stepan Rasp who figured out most of these tricks.
1. However, you define "sample", munge your data to have "sample" as its first dimension
2. Save the data as a zarr file chunked along a "sample" dimension with a fairly large chunk size. I would aim for 1 chunk to fill up about 75% of RAM.
2. True random data access is not necessary. It is usually good enough to shuffle within these chunks (not between them) provided they are large enough.

**Comment by tjvandal** [comment]
@nbren12  Thanks, I enjoyed your talk as well! The NOAA AI workshop was really interesting, looking forward to next years meeting. 

I can definitely see how using zarr would speed up access, I'll give that a try. For now, this doesn't solve the problem of making a second copy of the data but that may not be worth dealing with.

**Comment by nbren12** [comment]
Hey. I haven't decided to go again, but we should definitely meet in person if I do.

Unfortunately, I think copying the data in some way (either in memory or on disk) is necessary to ensure data locality and good performance since neural network training has a pretty unique access pattern.

**Comment by nbren12** [comment]
We can probably close this issue. Not sure I had many actionable thoughts above...
